precompiles/gmp/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+		let transfer_with_payload: WormholeTransferWithPayloadData = reader.read()?;
+
+		// get the wrapper for this asset by calling wrappedAsset()
+		// TODO: this should only be done if needed (when token chain == our chain)
+		let output = Self::call(
+			handle,
+			wormhole_bridge,
+			EvmDataWriter::new_with_selector(WRAPPED_ASSET_SELECTOR)
+				.write(transfer_with_payload.token_chain)
+				.write(transfer_with_payload.token_address)
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let wrapped_address: Address = reader.read()?;
+		log::debug!(target: "gmp-precompile", "wrapped token address: {:?}", wrapped_address);
+
+		// query our "before" balance (our being this precompile)
+		let output = Self::call(
+			handle,
+			wrapped_address.into(),
+			EvmDataWriter::new_with_selector(BALANCE_OF_SELECTOR)
+				.write(Address::from(handle.code_address()))
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let before_amount: U256 = reader.read()?;
+		log::debug!(target: "gmp-precompile", "before balance: {}", before_amount);
+
+		// our inner-most payload should be a VersionedUserAction
+		let user_action = VersionedUserAction::decode_with_depth_limit(
+			32,
+			&mut transfer_with_payload.payload.as_bytes(),
+		)
+		.map_err(|_| RevertReason::Custom("Invalid GMP Payload".into()))?;
+		log::debug!(target: "gmp-precompile", "user action: {:?}", user_action);
+
+		let currency_account_id = Runtime::AddressMapping::into_account_id(wrapped_address.into());
+
+		let currency_id: <Runtime as orml_xtokens::Config>::CurrencyId =
+			Runtime::account_to_currency_id(currency_account_id)
+				.ok_or(revert("Unsupported asset, not a valid currency id"))?;
+
+		let amount = transfer_with_payload
+			.amount
+			.try_into()
+			.map_err(|_| revert("Amount overflows balance"))?;
+
+		log::debug!(target: "gmp-precompile", "attempt to transfer {:?} of {:?}", amount, currency_id);
+
+		// Complete a "Contract Controlled Transfer" with the given Wormhole VAA.
+		// We need to invoke Wormhole's completeTransferWithPayload function, passing it the VAA.
+		// Upon success, it should have transferred tokens to this precompile's address.
+		Self::call(
+			handle,
+			wormhole_bridge,
+			EvmDataWriter::new_with_selector(COMPLETE_TRANSFER_WITH_PAYLOAD_SELECTOR)
+				.write(wormhole_vaa)
+				.build(),
+		)?;
+
+		// query our "after" balance (our being this precompile)
+		let output = Self::call(
+			handle,
+			wrapped_address.into(),
+			EvmDataWriter::new_with_selector(BALANCE_OF_SELECTOR)
+				.write(Address::from(handle.code_address()))
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let after_amount: U256 = reader.read()?;
+		log::debug!(target: "gmp-precompile", "after balance: {}", after_amount);
+
+		let amount_transfered = after_amount.saturating_sub(before_amount);
+
+		// TODO: review
+		if amount_transfered < transfer_with_payload.amount {
```

Comment:
Are you sure that this case can't be legit? It may depends on the wormhole's specifications, but the actual amount transferred may be lower for legitimate reasons (fees charged in the transferred token).

Reply:
I'm still skeptical, but from their contract code it looks like this should be the case.

Their fee mechanism:
1. only works on the native token (e.g. GLMR)
2. doesn't work for "type 3" payloads anyway (the kind we are using)

So it is likely that we will invent our own fee mechanism.

Anyway, I might change the logic for now:

```
let amount = amount_transferred.min(transfer_with_payload.amount);
```

Reply:
Changed in 773d44a, feel free to follow up or resolve.